File: typescript/src/p1.ts
function getInput(): string {
  return `forward 5
  down 5
  forward 8
  up 3
  down 8
  forward 2`
}

function parseLine(line: string): [number, number] {
  const [dir, a] = line.split(' ')
  const amount = +a // trick to convert to number
  let multiplier = 1
  let res: [number, number] = [0, 0]
  if (dir === 'up' || dir === 'backward') multiplier = -1
  if (dir === 'up' || dir === 'down') {
    res = [0, Number(amount) * multiplier]
  }
  if (dir === 'forward' || dir === 'backward') { 
    res = [Number(amount) * multiplier, 0]
  }

  return res
}

const items = getInput().split('\n').map(x => parseLine(x.trim()))
console.log(items)
const result = items.reduce((acc, ammount) => {
  acc[0] += ammount[0]
  acc[1] += ammount[1]
  return acc
}, [0, 0])

console.log(result, result[0] * result[1])

File: rust/src/bin/p1.rs
fn main() {
    let res = get_input()
        .lines()
        .map(parse_line)
        .fold(Point {x: 0, y: 0}, |mut acc, p| {
            acc.x += p.x;
            acc.y += p.y;
            return acc;
        });    

    println!("{:?}", res);
}

#[derive(Debug)]
struct Point {
    x: i32,
    y: i32,
}

fn parse_line(line: &'static str) -> Point { // we can also say &str.
    let (dir, ammount) = line.split_once(" ").expect("must contain a whitespace");
    let ammount = ammount.parse::<i32>().expect("must be an integer");
    // OR let ammount: i32 = ammount.parse().expect("must be an integer");
    // OR let ammount: i32 = str::parse(amount).expect("must be an integer");
    // There are more...
    if dir == "forward" {
        return Point { x: ammount, y: 0 };
    } else if dir == "down" {
        return Point { x: 0, y: ammount };
    } else {
        return Point { x: 0, y: -ammount};
    }
}

fn get_input() -> &'static str {
	return "forward 5
down 5
forward 8
up 3
down 8
forward 2"
}
